compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitBuffers.cpp

```diff
@@ -0,0 +1,145 @@
+  });
+
+  DenseSet<Operation *> toBeErased;
+  for (AMDAIE::DmaCpyNdOp l2ToL1DmaOp : l2ToL1DmaOps) {
+    LogicalObjectFifoFromMemrefOp sourceObjectFifo =
+        l2ToL1DmaOp.getSourceObjectFifo();
+    auto sourceAllocOp =
+        sourceObjectFifo.getMemref().getDefiningOp<memref::AllocOp>();
+    uint64_t sourceMemrefSpace = sourceObjectFifo.getMemorySpaceAsUInt();
+    if (!sourceAllocOp || sourceMemrefSpace != 1) continue;
+    LogicalObjectFifoFromMemrefOp targetObjectFifo =
+        l2ToL1DmaOp.getTargetObjectFifo();
+    Value targetAllocOp = targetObjectFifo.getMemref();
+
+    // Now we'll create a narrowed L2 buffer.
+    rewriter.setInsertionPoint(sourceAllocOp);
+    auto oldSourceMemRefType = cast<MemRefType>(sourceAllocOp.getType());
+    auto targetMemRefType = cast<MemRefType>(targetAllocOp.getType());
+    MemRefType newAllocType = MemRefType::get(
+        targetMemRefType.getNumElements(), targetMemRefType.getElementType(),
+        MemRefLayoutAttrInterface{}, oldSourceMemRefType.getMemorySpace());
+    auto newAllocOp = rewriter.create<memref::AllocOp>(rewriter.getUnknownLoc(),
+                                                       newAllocType);
+    auto newDeallocOp = rewriter.create<memref::DeallocOp>(
+        rewriter.getUnknownLoc(), newAllocOp);
+    newDeallocOp->moveBefore(&newAllocOp->getBlock()->back());
+
+    // Fetch the L3 -> L2 Dma Op corresponding to the L2 buffer as target.
+    AMDAIE::DmaCpyNdOp l3ToL2DmaOp;
+    for (Operation *objFifoUserOp : sourceObjectFifo->getUsers()) {
+      if (auto dmaOp = dyn_cast<AMDAIE::DmaCpyNdOp>(objFifoUserOp);
+          dmaOp.getTargetObjectFifo() == sourceObjectFifo) {
+        l3ToL2DmaOp = dmaOp;
+        toBeErased.insert(dmaOp);
+        break;
+      }
+    }
+    toBeErased.insert(sourceAllocOp);
+    toBeErased.insert(sourceObjectFifo);
+
+    auto type = cast<MemRefType>(newAllocOp.getType());
+    // Create new logicalobjectfifo.from_memref for the newly created L2 buffer.
+    rewriter.setInsertionPoint(l2ToL1DmaOp.getSourceObjectFifo());
+    auto source = rewriter.create<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+        rewriter.getUnknownLoc(), LogicalObjectFifoType::get(type),
+        newAllocOp.getResult(), sourceObjectFifo.getTiles());
+
+    // Create new L3 -> L2 Dma Op.
+    rewriter.setInsertionPoint(l3ToL2DmaOp);
+    rewriter.create<AMDAIE::DmaCpyNdOp>(
+        l3ToL2DmaOp.getLoc(), source, l3ToL2DmaOp.getTargetMixedOffsets(),
+        l3ToL2DmaOp.getTargetMixedSizes(), l3ToL2DmaOp.getTargetMixedStrides(),
+        l3ToL2DmaOp.getSource(), l3ToL2DmaOp.getSourceMixedOffsets(),
+        l3ToL2DmaOp.getSourceMixedSizes(), l3ToL2DmaOp.getSourceMixedStrides());
+
+    // Create new L2 -> L1 Input DmaOp.
+    rewriter.setInsertionPoint(l2ToL1DmaOp);
+    auto newL2ToL1DmaOp = rewriter.create<AMDAIE::DmaCpyNdOp>(
+        l2ToL1DmaOp.getLoc(), l2ToL1DmaOp.getTarget(),
+        l2ToL1DmaOp.getTargetMixedOffsets(), l2ToL1DmaOp.getTargetMixedSizes(),
```

Comment:
The logic here doesn't take changes offsets into account and these might need to change. For example, dma1: [0] [16] [1], dma2: [16] [16] [1]. After splitting, the dma2 offset needs to become zero to be correct. I think more test cases are needed + think about complex cases, overlapping access patterns for example.

Reply:
Good catch!

Here's what I'm currently doing : Since the new L2 buffer we're creating is linearized form of the L1 buffer's shape, I'm creating new offset/size/stride as per this linearized shape.
```
Eg: L1: <1x1x8x4x8x4>
    L2: <1024> (we get this by linearizing L1's buffer : 1x1x8x4x8x4 = 1024)
      
    So, because there is now 1:1 correspondence between L1 and L2
    I'm creating the offset/size/stride as [0,0,0,0][1,1,1,1024][0,0,1024,1]
      
    And since the same L2 buffer will be fed from L3, I'm using the same access pattern as above.
```

I might be wrong here but this linearized access pattern should also take care of overlapping access patterns ? Because of the 1:1 correspondence between L1 and L2 ?

Could you check and let me know if it looks okay to you?

Reply:
> Since the new L2 buffer we're creating is linearized form of the L1 buffer's shape, I'm creating new offset/size/stride as per this linearized shape.

That's not correct as changing the offset/stride/access pattern changes the data packing and will therefore result in incorrect results.

> I might be wrong here but this linearized access pattern should also take care of overlapping access patterns ?

Were do you see an issue with overlapping access patterns?